### PR TITLE
Add support for B2C authentication

### PR DESCRIPTION
**Pull request by ReeceHoffmann**
Ready for full review.

Adds b2c support on the front-end. (This configuration of B2C mode is not supported by the backend. Client is fully compatible with backend running in regular auth mode)

Screenshot of login page:
image

### CODE REVIEW

File: run.js [hunk 1]
```diff
@@ -32,6 +33,31 @@ program
...
+    "Name of b2c login userflow",
+    process.env.VT_UI_B2C_USERFLOW || ""
+  )
+  .option(
+    "--b2c-tenant <b2c_tenant>",
+    "Name of b2c tenant",
+    process.env.VT_UI_B2C_TENANT || ""
+  )
+  .option(
+    "--b2c-clientId <b2c_clientId>",
```
Comment (igboyes) [inline run.js]:
```suggestion
    "--b2c-client-id <b2c_client_id>",
```

File: run.js [hunk 2]
```diff
@@ -32,6 +33,31 @@ program
...
+    "The frontend b2c application clientId",
+    process.env.VT_UI_B2C_CLIENTID || ""
```
Comment (igboyes) [inline run.js]:
```suggestion
    process.env.VT_UI_B2C_CLIENT_ID || ""
```

File: run.js [hunk 3]
```diff
@@ -32,6 +33,31 @@ program
...
+  )
+  .option(
+    "--b2c-scope <b2c_scope>",
+    "The B2C backend api scope",
+    process.env.VT_UI_SCOPE || ""
```
Comment (igboyes) [inline run.js]:
```suggestion
    process.env.VT_UI_B2C_SCOPE || ""
```

File: server/csp.js [hunk 4]
```diff
@@ -4,16 +4,22 @@ const fontAwesomeURL = "https://use.fontawesome.com";
 const defaultSrc = "default-src 'self'";
 const imgSrc = "img-src 'self' data:";
 const fontSrc = `font-src 'self' ${fontAwesomeURL}`;
-const connectSrc = "connect-src 'self' sentry.io";
+const connectSrc = "connect-src 'self' sentry.io https://to9001.b2clogin.com";
```
Comment (igboyes) [inline server/csp.js]: Should this be hard-coded?

One thought. Is it possible to use custom domains with B2C (eg. b2c.virtool.ca). In that case, this URL would also need to be configurable.

File: server/csp.js [hunk 5]
```diff
@@ -4,16 +4,22 @@ const fontAwesomeURL = "https://use.fontawesome.com";
...
 
 const generateCSPScriptSrc = (nonce) => {
-  return `script-src 'self' 'nonce-${nonce}' ${fontAwesomeURL}`;
+  return `script-src 'self' 'nonce-${nonce}' ${fontAwesomeURL} `;
```
Comment (igboyes) [inline server/csp.js]: Intentional?

File: server/templateOptions.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  if (
+    options.useB2c &&
+    !(
+      options.b2cUserflow &&
+      options.b2cTenant &&
+      options.b2cClientId &&
+      options.b2cScope
+    )
+  ) {
+    console.log(
```
Comment (igboyes) [inline server/templateOptions.js]: `console.error` may make more sense here.

File: server/templateOptions.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    process.exit(0);
+  }
+
+  return {
+    use_b2c: options.useB2c,
+    b2c_userflow: options.b2cUserflow,
+    b2c_tenant: options.b2cTenant,
+    b2c_clientId: options.b2cClientId,
+    b2c_scope: options.b2cScope,
+    sentryDSN: options.sentryDsn,
```
Comment (igboyes) [inline server/templateOptions.js]: May as well be consistent with casing here (`sentry_dsn`).

File: src/js/app/authConfig.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+                }
+            }
+        }
+    };
+};
+
+export const getProtectedResources = () => ({
+    backendApi: {
+        scopes: [
+            `https://${window.b2c.tenant}.onmicrosoft.com/0c424b39-4b26-42d7-a494-f393addf6a01/${window.b2c.scope}`
```
Comment (igboyes) [inline src/js/app/authConfig.js]: Isn't the string here sensitive or not applicable to all deployments?

File: src/js/index.js [hunk 9]
```diff
@@ -8,19 +8,25 @@ import App from "./app/App";
...
+        use: true
+    };
 }
 
 Sentry.init({
     dsn: window.virtool.sentryDSN
 });
 
 const sentryCheck = res => {
+    console.log(res.body);
```
Comment (igboyes) [inline src/js/index.js]:
```suggestion
```

File: src/js/samples/sagas.js [hunk 10]
```diff
@@ -91,7 +91,7 @@ export function* findReadFiles() {
 
 export function* getSample(action) {
     try {
-        const response = yield samplesAPI.get(action.payload);
+        const response = yield callWithAuthentication(samplesAPI.get, action.payload);
```
Comment (igboyes) [inline src/js/samples/sagas.js:94]: Is this for debugging?

File: src/js/labels/sagas.js [hunk 11]
```diff
@@ -13,8 +13,7 @@ export function* watchLabels() {
 }
 
 export function* listLabels(action) {
-    const response = yield labelsAPI.listLabels(action);
-    yield put(listLabelsSuccess(response.body));
+    yield apiCall(labelsAPI.listLabels, action.payload, LIST_LABELS);
```
Comment (igboyes) [inline src/js/labels/sagas.js:16]: Why changed? If unrelated to B2C, it should be dropped from PR.

File: src/js/utils/sagas.js [hunk 12]
```diff
@@ -46,6 +48,43 @@ export function* apiCall(apiMethod, action, actionType, context = {}) {
...
 }
 
+export function* callWithAuthentication(apiMethod, action) {
+    if (window.b2c.use) {
+        const accessToken = yield getAccessToken();
+        return yield apiMethod(action).set("Bearer", accessToken);
+    } else {
+        return yield apiMethod(action);
+    }
+}
```
Comment (igboyes) [inline src/js/utils/sagas.js]:
```suggestion
    if (window.b2c.use) {
        const accessToken = yield getAccessToken();
        return yield apiMethod(action).set("Bearer", accessToken);
    }
    
    return yield apiMethod(action);
}
```

File: src/js/utils/sagas.js [hunk 13]
```diff
@@ -46,6 +48,43 @@ export function* apiCall(apiMethod, action, actionType, context = {}) {
...
+
+export function getAccessToken() {
+    const protectedResources = getProtectedResources();
+    console.log("activeAccount", msalInstance.getActiveAccount());
+    console.log(msalInstance.getAllAccounts());
+    console.log(protectedResources.backendApi.scopes);
```
Comment (igboyes) [inline src/js/utils/sagas.js]: Remove before merge.
```suggestion
```

**Review by igboyes — CHANGES_REQUESTED** [review]

File: server/csp.js (hunk 4 above)
Comment (ReeceHoffmann) [reply server/csp.js]: Ah oops this line should have been deleted, generating a dynamic connect-src is done by the function `generateCSPConnectSrc` regardless of B2C status. Line will be deleted in next push.

File: server/csp.js (hunk 5 above)
Comment (ReeceHoffmann) [reply server/csp.js]: No, changed by accident, removed.

File: src/js/app/authConfig.js (hunk 8 above)
Comment (ReeceHoffmann) [reply src/js/app/authConfig.js]: Fortunately not sensitive (client id of the backend) but it is deployment specific. Changed to be specified by front-end CLI

**Review by igboyes — APPROVED** [review]

**Pull request merged**